### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- 3.3.2
- 2.1.4
- 1.2.3

#### ANS  

**Answer:** 3.3.2

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>


- 2MB
- 24MB
- 100MB
- 250MB

In [98]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz

--2023-03-03 16:11:52--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 140.82.113.4
Connecting to github.com (github.com)|140.82.113.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230303%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230303T161152Z&X-Amz-Expires=300&X-Amz-Signature=712d2ae0f100c7a094ea85f62950a7d24e554be7c4fb4484ce8cd7d8cdcda190&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-03 16:11:52--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e

In [96]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()


fhv_df = spark.read \
        .option("header", "true") \
        .csv('fhvhv_tripdata_2021-06.csv.gz')

fhv_df.repartition(12).write.parquet('fhvhv_2021_06_repartitioned')

In [97]:
!ls -lh fhvhv_2021_06_repartitioned

total 297M
-rw-r--r-- 1 mmm mmm   0 Mar  3 16:07 _SUCCESS
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00000-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00001-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00002-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00003-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00004-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00005-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00006-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00007-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.snappy.parquet
-rw-r--r-- 1 mmm mmm 25M Mar  3 16:07 part-00008-c4d8b388-1755-4b2e-af44-35ba62680db9-c000.sna

### ANS

**Answer:** 24 MB (closest answer).

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- 452,470
- 50,982

In [99]:
from pyspark.sql.types import TimestampType

fhv_df = fhv_df.withColumn('pickup_datetime', fhv_df.pickup_datetime.cast(TimestampType()))
fhv_df = fhv_df.withColumn('dropoff_datetime', fhv_df.dropoff_datetime.cast(TimestampType()))

In [100]:
fhv_df.withColumn('pickup_date', F.to_date('pickup_datetime')) \
    .filter('pickup_date = "2021-06-15"') \
    .count()

452470

#### ANS

**Answer:** 452470

### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- 66.87 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [102]:
from pyspark.sql.functions import asc, desc
from pyspark.sql import functions as F

fhv_df \
    .withColumn('trip_duration_in_sec', F.col('dropoff_datetime').cast('long') - F.col('pickup_datetime').cast('long')) \
    .withColumn("trip_duration_in_hrs", F.round(F.col("trip_duration_in_sec")/3600,2)) \
    .sort(F.col('trip_duration_in_hrs').desc()) \
    .show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+--------------------+--------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|trip_duration_in_sec|trip_duration_in_hrs|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+--------------------+--------------------+
|              B02872|2021-06-25 13:55:41|2021-06-28 08:48:25|          98|         265|      N|                B02872|              240764|               66.88|
|              B02765|2021-06-22 12:09:45|2021-06-23 13:42:44|         188|         198|      N|                B02765|               91979|               25.55|
|              B02879|2021-06-27 10:32:29|2021-06-28 06:31:20|          78|         169|      N|                B02879|               71931|               19.98|
|              B02800|2021-0

#### ANS

**Answer:** 66.87 Hours (closest answer).

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- 4040
- 8080

#### ANS

**Answer:** local port: 4040

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- Crown Heights North

In [103]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2023-03-03 16:20:37--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230303%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230303T162037Z&X-Amz-Expires=300&X-Amz-Signature=38dd8b4882ffea1770d804166afaf3b520b2624e29a08d2ac366c3a6d872c8b9&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-03 16:20:37--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [81]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

zone_df = spark.read \
    .option("header", "true") \
    .csv('taxi_zone_lookup.csv')

zone_df.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [94]:
fhv_df_pl_grouped = fhv_df.groupBy('PULocationID') \
    .count() \
    .sort(F.col('count').desc())

In [95]:
fhv_df_pl_grouped.join(zone_df, fhv_df_pl_grouped.PULocationID == zone_df.LocationID, how='left') \
                    .show()

+------------+------+----------+---------+--------------------+------------+
|PULocationID| count|LocationID|  Borough|                Zone|service_zone|
+------------+------+----------+---------+--------------------+------------+
|          61|231279|        61| Brooklyn| Crown Heights North|   Boro Zone|
|          79|221244|        79|Manhattan|        East Village| Yellow Zone|
|         132|188867|       132|   Queens|         JFK Airport|    Airports|
|          37|187929|        37| Brooklyn|      Bushwick South|   Boro Zone|
|          76|186780|        76| Brooklyn|       East New York|   Boro Zone|
|         231|164344|       231|Manhattan|TriBeCa/Civic Center| Yellow Zone|
|         138|161596|       138|   Queens|   LaGuardia Airport|    Airports|
|         234|158937|       234|Manhattan|            Union Sq| Yellow Zone|
|         249|154698|       249|Manhattan|        West Village| Yellow Zone|
|           7|152493|         7|   Queens|             Astoria|   Boro Zone|

#### ANS

**Answer:**  Crown Heights North